### DISCUSSION TITLE
How to use custom actions for use with ModbusSimulatorContext?

### DISCUSSION THREAD

**Discussion by benjaminorr** [pymodbus-dev/pymodbus#1465] [Help] [answered]
I'm working on a server simulator application with aspects that closely resemble [this example](https://github.com/pymodbus-dev/pymodbus/blob/08845c7b65d05c8d0090aa774e79900e6358a6c3/examples/server_simulator.py). 

The device I'm simulating has defined minimum and maximum ranges for each register. This is independent of the type (uint16, etc). In order to restrict the client from writing values beyond these ranges, I'd like to use custom actions (as mentioned in the docs and above example) to do this, however I can't find a good example or explanation of how to do so. 

I'm configuring my datastore with a `.json5` file that defines ranges and values, etc as seen in the example.

Any help on how to do this via custom actions or other methods would be super appreciated (especially working code examples!).

Thanks

**Comment by janiversen** [comment, marked as answer]
look at the supplied example.

**Reply by janiversen** [reply]
ahh I see you found our example! then all you need to do is adapt the configuration to your needs and of course write the action.

Be aware though as pr documentation actions are there to change the registers not to verify the them. If you want an extended verification, the simplest solution is to inherit the simulator context class and overwrite the verification method.
